packages/clay-form/src/InputWithAutocomplete.tsx

```diff
@@ -0,0 +1,102 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import ClayAutocomplete from '@clayui/autocomplete';
+import ClayDropDown from '@clayui/drop-down';
+import React, {useEffect, useRef, useState} from 'react';
+import {useKeyHandlerForList} from '@clayui/shared';
+
+interface IProps extends React.InputHTMLAttributes<HTMLInputElement> {
+	/**
+	 * Flag to indicate is loading indicator should be shown.
+	 */
+	loading?: boolean;
+
+	/**
+	 * Items that will show up in the dropdown list
+	 */
+	items: Array<string> | Array<Object>;
+
+	/**
+	 * Function that returns the value that should be showed in the list.
+	 */
+	itemSelector?: (val: string | Object) => string;
+
+	/**
+	 * Callback for when input value changes
+	 */
+	onItemSelect: (val: string | Object) => string | Object;
+
+	/**
+	 * Callback for when input value changes
+	 */
+	onValueChange: (val: string | Object) => void;
+}
+
+export const ClayInputWithAutocomplete: React.FunctionComponent<IProps> = ({
+	items = [],
+	itemSelector = val => val as string,
+	loading,
+	onItemSelect,
+	onValueChange,
+	value = '',
+	...otherProps
+}) => {
+	const activeListItemRef = useRef<HTMLLIElement>(null);
+	const inputRef = useRef<HTMLInputElement>(null);
+	const [activeIndex, setActiveIndex] = useState(-1);
+	const [active, setActive] = useState(!!value);
+
+	const keyHandler = useKeyHandlerForList({
+		activeListItemRef,
+		index: activeIndex,
+		inputRef,
+		onIndexChange: setActiveIndex,
+		onIndexSelect: (index: number) => onItemSelect(items[index]),
+		totalItems: items.length,
+	});
+
+	useEffect(() => {
+		setActive(!!value);
+	}, [value]);
+
+	useEffect(() => {
+		setActiveIndex(-1);
+	}, [active]);
+
+	return (
+		<ClayAutocomplete>
+			<ClayAutocomplete.Input
+				{...otherProps}
+				onChange={e => onValueChange(e.target.value)}
+				onKeyDown={keyHandler}
+				ref={inputRef}
+				value={value}
+			/>
+
+			<ClayAutocomplete.DropDown active={active} onSetActive={setActive}>
+				<ClayDropDown.ItemList>
+					{items.map((item, i) => (
+						<ClayAutocomplete.Item
```

Comment:
I was going to say to make it extensible but I remembered that we have a speech at #2152, so I think we can add that later 🙂.